### ISSUE TITLE
 False positive with default arg values before *args

### ISSUE THREAD

**Issue by jaraco** [PyCQA/pylint#7396]
This bug is a re-file of #2481. That issue was locked without resolution.

The issue reported there is still an issue, and is affecting code [like this one](https://foss.heptapod.net/mercurial/mercurial-devel/-/blob/f59e49f6bee4591763ccdad9cf0d7fb2c5b130c7/mercurial/commands.py#L4093).

The problem with W1113 is that it is inconsiderate of the case where a positional argument has a default value and it's not expected that argument would ever be supplied in keyword arguments.

The suggestion that the positional arg with a default should be placed after the varargs is incorrect. It violates the intended usage (where the default param is meant to be the optional first of a variable number of args).

Even without PEP 570, it's perfectly legitimate for a function to accept default values for parameters that are meant to be passed positionally.

In my opinion, a linter shouldn't emit warnings for code that has legitimate use even if it could potentially be misused in select cases, but especially in the PEP 570 case, there's no potential for mis-use, so there's no value in the spurious warning.

Yes it's true, one would get an error to call it as a keyboard parameter and a positional parameter, but the same is true for any function:

```
>>> def f(param1=None):
...   pass
... 
>>> f('a', param1='b')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: f() got multiple values for argument 'param1'
```

That's no justification for warning on `def f(param1)` or `def f(param1=None)`.

The same principle applies whether or not there are varargs present.

W1113 should be removed or disabled by default (opt-in only).

**Comment by jaraco** [comment]
Regarding "Needs specification", would it be sufficient to say "W1113 should be disabled by default" as the specification? What more would need to be specified to move forward with an implementation?

**Comment by Pierre-Sassoulas** [comment]
What I meant by need specifications is that we need to define what we want for the test cases in https://github.com/PyCQA/pylint/blob/main/tests/functional/k/keyword_arg_before_vararg.py. I think the check can be modified to reduce false positives by checking if the positional argument has a default value. But you seem to think the check can't be salvaged ? (It's possible to disable it yourself in the configuration).

**Comment by jaraco** [comment]
In my opinion, those test cases are fine for illustrating the condition the check intends to catch, and maybe the tests just need to be extended to illustrate that the check is one that is not enabled by default.

I think part of my problem is that I'm biased. I've only ever encountered this lint warning twice (in the above referenced code and a similar derivative case), so in all cases, I've only ever experienced the false positive. I think it would be helpful to see an example where this lint warning is useful.

The only example I saw was at [this page](https://vald-phoenix.github.io/pylint-errors/plerr/errors/typecheck/W1113.html), which clearly breaks the meaning of the code and disallows `y` to be passed positionally where it was before (and also breaks `x` incidentally).

I looked through the history to try to understand better how this warning is intended to work, what it's protecting. I found more confusion in #2027 and #1835, but no documentation on why this form is discouraged. Why not assume that the user wrote what the signature suggests (a positional argument with a default value) and not assume they mistakenly meant for that to be a keyword-only argument? Is there any other background on why this warning was created?

**Comment by DanielNoord** [comment]
I think this is a leftover from the same paradigm that influenced the `star-args` message, see https://stackoverflow.com/questions/28915233/alternative-to-asterisk-in-function-definitions and some discussion there.

I wasn't involved then, but `pylint` maintainers at the time seemed to discourage using star args and anything related to them. I think we might want to consider removing this message, it does indeed seem to serve no real purpose? This is perfectly valid code and the "Good code" example in our documentation looks very unpythonic to me.

_In fact, I would argue a `put-args-or-kwargs-at-the-end-of-the-signature` message makes more sense. The name needs some work though..._

**Comment by Pierre-Sassoulas** [comment]
> The name needs some work though...

``misplaced-variable-length-args`` ?

**Comment by DanielNoord** [comment]
Since this is probably controversial to some and mostly stylistic, what about:
`consider-placing-variable-length-args-last`? The message is a bit long but it is also actionable, which might help users.

**Comment by Pierre-Sassoulas** [comment]
Ok, so in order to fix this issue we should create a new extension for ``consider-placing-variable-length-args-last`` and refactor the existing check to become an optional extension too.

*Off topic: I feel like we need to relax the prefix for checker requirement in message id first, because blocking 99 msgid for this seems inefficient.*

**Comment by DanielNoord** [comment]
I think we don't need an extension but we can put it under `CodeStyle`.

I'm not sure refactoring is worth it. I really can't see why anybody would need this check and think it is actually bad practice to write code like in the "Good example". I would be in favour of removing just like we have done with `star-args`.

**Comment by Pierre-Sassoulas** [comment]
> I think we don't need an extension but we can put it under CodeStyle.

This is probably something we need to discuss but the more I think about it the more I think we should separate checkers **unless there's actually checks that would be duplicated and grouping those checks together make performance better**.  It's more modular, it makes for simpler checkers, the file are shorter, the checker's owner is obvious and can be defined in the CODEOWNER file, it's impossible to shoot oneself in the foot with the ``only_required_for_check`` decorator...

>  I would be in favour of removing just like we have done with star-args.

It seems like this is controversial reading the opinion of implementers in the past. Wouldn't making it optional be more prudent ?

**Comment by DanielNoord** [comment]
> This is probably something we need to discuss but the more I think about it the more I think we should separate checkers **unless there's actually checks that would be duplicated and grouping those checks together make performance better**. It's more modular, it makes for simpler checkers, the file are shorter, the checker's owner is obvious and can be defined in the CODEOWNER file, it's impossible to shoot oneself in the foot with the `only_required_for_check` decorator...

I have actually been thinking the opposite 😅 The `visit_` pattern with individual checker classes per check creates a lot of overhead, especially with `safe_infer` being called everywhere. There is a very high likelihood that we are not hitting caches across multiple `visit_` calls to the same node type I think.

> It seems like this is controversial reading the opinion of implementers in the past. Wouldn't making it optional be more prudent ?

I mean, for `*args` it makes no sense as there is no risk of "taking over" the keyword argument as discussed here:
https://github.com/PyCQA/pylint/issues/2027
For `**kwargs` `def g(**kwargs, x=1):` is invalid syntax.

If anything this is probably more confusing than following the standard pattern of just putting them at the end.

**Comment by Pierre-Sassoulas** [comment]
Blocked by #6670. The message is documented with bad/good and we want to keep that documentation for users using the old versions (where the message will still exists). There's nothing in place to do that right now.
